Doctrine/Orm/Extension/PaginationExtension.php

```diff
@@ -129,4 +142,229 @@ private function getItemsPerPage(ResourceInterface $resource, Request $request)
+                            $parentAlias = substr($relationship, 0, $pos);
+                            $association = substr($relationship, $pos + 1);
+
+                            $parentMetadata = $this->getClassMetadataFromJoinAlias($parentAlias, $queryBuilder);
+
+                            if ($parentMetadata->isCollectionValuedAssociation($association)) {
+                                $useOutputWalkers = true;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        return $useOutputWalkers;
+    }
+
+    /**
+     * Gets the class metadata from a given join alias.
+     *
+     * @param string       $alias
+     * @param QueryBuilder $queryBuilder
+     *
+     * @return ClassMetadata
+     */
+    private function getClassMetadataFromJoinAlias($alias, QueryBuilder $queryBuilder)
+    {
+        $rootEntities = $queryBuilder->getRootEntities();
+        $rootAliases = $queryBuilder->getRootAliases();
+
+        $joinParts = $queryBuilder->getDQLPart('join');
+
+        $aliasMap = [];
+        $targetAlias = $alias;
+
+        foreach ($joinParts as $rootAlias => $joins) {
+            $aliasMap[$rootAlias] = 'root';
+
+            foreach ($joins as $join) {
+                $alias = self::getJoinAlias($join);
+                $relationship = self::getJoinRelationship($join);
+
+                $pos = strpos($relationship, '.');
+
+                $aliasMap[$alias] = [
+                    'parentAlias' => substr($relationship, 0, $pos),
+                    'association' => substr($relationship, $pos + 1),
+                ];
+            }
+        }
+
+        $associationStack = [];
+        $rootAlias = null;
+
+        while (null === $rootAlias) {
+            $mapping = $aliasMap[$targetAlias];
+
+            if ('root' === $mapping) {
+                $rootAlias = $targetAlias;
+            } else {
+                $associationStack[] = $mapping['association'];
+                $targetAlias = $mapping['parentAlias'];
+            }
+        }
+
+        $rootEntity = $rootEntities[array_search($rootAlias, $rootAliases)];
+
+        $rootMetadata = $this
+            ->managerRegistry
+            ->getManagerForClass($rootEntity)
+            ->getClassMetadata($rootEntity)
+        ;
+
+        $metadata = $rootMetadata;
+
+        while (null !== ($association = array_pop($associationStack))) {
+            $associationClass = $metadata->getAssociationTargetClass($association);
+
+            $metadata = $this
+                ->managerRegistry
+                ->getManagerForClass($associationClass)
+                ->getClassMetadata($associationClass)
+            ;
+        }
+
+        return $metadata;
+    }
+
+    /**
+     * Gets the relationship from a Join expression.
+     *
+     * @param Join $join
+     *
+     * @return string
+     */
+    private static function getJoinRelationship(Join $join)
```

Comment:
Instead of using these methods, you can just use a decorator or an adapter for this `Join` object and then be sure that the `getJoin` method is always there.